Firestore/Example/Firestore_Benchmarks_iOS/FSTLevelDBBenchmarkTests.mm

```diff
@@ -0,0 +1,180 @@
+#include "Firestore/core/src/firebase/firestore/local/leveldb_transaction.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+using firebase::firestore::local::LevelDbTransaction;
+using firebase::firestore::model::DatabaseId;
+
+// Pre-existing document size
+static const int kDocumentSize = 1024 * 2;  // 2 kb
+
+static std::string DocumentData() {
+  return std::string(kDocumentSize, 'a');
+}
+
+static std::string UpdatedDocumentData(int documentSize) {
+  return std::string(documentSize, 'b');
+}
+
+static NSString *LevelDBDir() {
+  NSError *error;
+  NSFileManager *files = [NSFileManager defaultManager];
+  NSString *dir =
+      [NSTemporaryDirectory() stringByAppendingPathComponent:@"FSTPersistenceTestHelpers"];
+  if ([files fileExistsAtPath:dir]) {
+    // Delete the directory first to ensure isolation between runs.
+    BOOL success = [files removeItemAtPath:dir error:&error];
+    if (!success) {
+      [NSException raise:NSInternalInconsistencyException
+                  format:@"Failed to clean up leveldb path %@: %@", dir, error];
+    }
+  }
+  return dir;
+}
+
+static FSTLevelDB *LevelDBPersistence() {
+  // This owns the DatabaseIds since we do not have FirestoreClient instance to own them.
+  static DatabaseId database_id{"p", "d"};
+
+  NSString *dir = LevelDBDir();
+  FSTSerializerBeta *remoteSerializer = [[FSTSerializerBeta alloc] initWithDatabaseID:&database_id];
+  FSTLocalSerializer *serializer =
+      [[FSTLocalSerializer alloc] initWithRemoteSerializer:remoteSerializer];
+  FSTLevelDB *db = [[FSTLevelDB alloc] initWithDirectory:dir serializer:serializer];
+  NSError *error;
+  BOOL success = [db start:&error];
+  if (!success) {
+    [NSException raise:NSInternalInconsistencyException
+                format:@"Failed to create leveldb path %@: %@", dir, error];
+  }
+
+  return db;
+}
+
+class LevelDBFixture : public benchmark::Fixture {
+  virtual void SetUp(benchmark::State &state) {
+    _db = LevelDBPersistence();
+    FillDB();
+  }
+
+  virtual void TearDown(benchmark::State &state) {
+    _db = nil;
+  }
+
+  void FillDB() {
+    LevelDbTransaction txn(_db.ptr, "benchmark");
+
+    for (int i = 0; i < _numDocuments; i++) {
+      FSTDocumentKey *docKey =
+          [FSTDocumentKey keyWithPathString:[NSString stringWithFormat:@"docs/doc_%i", i]];
+      std::string docKeyString = [FSTLevelDBRemoteDocumentKey keyWithDocumentKey:docKey];
+      txn.Put(docKeyString, DocumentData());
+      WriteIndex(txn, docKey);
+    }
+    txn.Commit();
+    // Force a write to disk to simulate startup situation
+    _db.ptr->CompactRange(NULL, NULL);
+  }
+
+ protected:
+  void WriteIndex(LevelDbTransaction &txn, FSTDocumentKey *docKey) {
+    txn.Put([FSTLevelDBDocumentTargetKey keyWithDocumentKey:docKey targetID:_targetID],
+            _emptyBuffer);
+    txn.Put([FSTLevelDBTargetDocumentKey keyWithTargetID:_targetID documentKey:docKey],
+            _emptyBuffer);
+  }
+
+  FSTLevelDB *_db;
```

Comment:
Optional nit: I don't know the rules for mixed code, but C++ style guide prescribes trailing, not leading, underscores for private variables.

Reply:
Done.